src/lib.rs

```diff
@@ -1,59 +1,74 @@
-use serde_json::{Value};
+use serde_json::{Map, Value};
+use thiserror::Error;
 
-pub fn hydrate(base: Value, item: &mut Value) {
-    if item.is_null(){
-        *item = base.clone();
-    }
-    else if item.is_object() {
-        if item.as_object() == base.as_object(){
-            return;
+const MAGIC_MARKER: &str = "𒍟※";
+
+pub trait Hydrate {
+    fn hydrate(&mut self, base: Self) -> Result<(), Error>;
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("type mismatch")]
+    TypeMismatch(Value, Value),
+}
+
+impl Hydrate for Value {
+    fn hydrate(&mut self, base: Self) -> Result<(), Error> {
+        match self {
+            Value::Object(item) => match base {
+                Value::Object(base) => item.hydrate(base),
+                _ => Err(Error::TypeMismatch(self.clone(), base)),
+            },
+            Value::Array(item) => match base {
+                Value::Array(base) => item.hydrate(base),
+                _ => Err(Error::TypeMismatch(self.clone(), base)),
+            },
+            _ => Ok(()),
```

Comment:
Do we want to ensure the base is the same type as the item here? Or do we not care because there's nothing we can do, hydration-wise?

Reply:
If the item is the magic marker, then we don't want to check if things are the same type. Otherwise, I think that if there is a type mismatch we should either let the item value trump or throw an error (perhaps this should be a flag of how strict it should be).